### PR TITLE
Add Spanner Plugin

### PR DESCRIPTION
**Pull request by sandeepkumarverma1**
Signed-off-by: Sandeep Kumar Verma1 <sandeep.kumar.verma1@ibm.com>

### CODE REVIEW

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/sample-audit-logs.json [hunk 1]
```diff
@@ -0,0 +1,235 @@
+[
```
Comment (PratikshaSonawane) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/sample-audit-logs.json:1]: We can remove this file, to keep it consistent with other plugins structure

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+- This is a Logstash filter plug-in for the universal connector that is featured in IBM Security Guardium. It parses events and messages from the Spanner audit logs into a Guardium record instance 
+
+```
+filter {
+if [type] == "spanner_db" {
+	spanner_db_guardium_filter {}
+  }
+}
+
+```
```
Comment (Tal-Daniel) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md:206]: Please shorten by linking to conf and referring to copy what's within the filter section there, instead of copying the conf here. Refer to other plug-ins.

Also, the sentence feels like the first sentence in the README, rather than a sentence dedicated to this stage. Missing instruction to paste configuration in the Guardium data protection relevant place in the UI.

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
...tionInfo':[{'resource':'projects\/project-sccd\/instances\/spanner-test\/databases\/test','permission':'spanner.databases.select','granted':true,'resourceAttributes':{'service':'spanner','name':'projects\/project-sccd\/instances\/spanner-test\/databases\/test','type':'spanner.databases'}}],'resourceName':'projects\/project-sccd\/instances\/spanner-test\/databases\/test\/sessions\/AH-i_0ZRnChs6K4dbngNt3ICWHWlM7PEK_0zGralREXja09C-tbna2rqiPl9yQ','request':{'queryMode':'PROFILE','@type':'type.googleapis.com\/google.spanner.v1.ExecuteSqlRequest','session':'projects\/project-sccd\/instances\/spanner-test\/databases\/test\/sessions\/AH-i_0ZRnChs6K4dbngNt3ICWHWlM7PEK_0zGralREXja09C-tbna2rqiPl9yQ','sql':'select * from singers'}},'insertId':'1lpovsaf2u0n15','resource':{'type':'spanner_instance','labels':{'instance_id':'spanner-test','project_id':'project-sccd','instance_config':'','location':'us-central1'}},'timestamp':'2022-04-26T12:02:34.852924678Z','severity':'INFO','logName':'projects\/project-sccd\/logs\/cloudaudit.googleapis.com%2Fdata_access','operation':{'id':'12631762209399029108','producer':'spanner.googleapis.com','first':true},'receiveTimestamp':'2022-04-26T12:02:36.143153614Z'}"
```
Comment (Tal-Daniel) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md:216]: same goes here: refer and link to example conf for testing, for developers only, ... 

@AlexaNahum

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+}
+
+output {
+        stdout { codec => rubydebug }
+}
+
+
+```
+## Limitations & Assumptions
+1. Error Logs are not generated in GCP for spanner and this plugin does not support errors traffic in guardium.
```
Comment (Tal-Daniel) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md:234]: Spanner

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+2. Audit/Data access log doesn't contains server ip, Default value is set 0.0.0.0 as server IP.
+3. Some fields could not be mapped as no information regarding the same was available in the logs. Following important fields couldn't be mapped-
+    - Source program : Not available with logs
+    - OS User : Not available with logs
+    - Client HostName : Not available with logs
+4. For Drop Database there is no DDL query required in spanner as there is only Delete button And on deleting database through spanner UI, we are not getting query parameter in audit logs. please refer [spanner DDL docs link](https://cloud.google.com/spanner/docs/reference/standard-sql/data-definition-language) for your reference. Hereforth,this is not captured in fullSql report.
```
Comment (Tal-Daniel) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md:240]: "...[, a]nd on deleting"
@AlexaNahum

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+
+
+## Configuring the Spanner filter in Guardium
+The Guardium universal connector is the Guardium entry point for native audit/data_access logs. The Guardium universal connector identifies and parses the received events, and converts them to a standard Guardium format. The output of the Guardium universal connector is forwarded to the Guardium sniffer on the collector, for policy and auditing enforcements. Configure Guardium to read the native audit/data_access logs by customizing the Spanner template.
+
+## Before you begin
+You must have permission for the S-Tap Management role. The admin user includes this role by default.
+Download the guardium_logstash-offline-plugin-gcp-pubsub-spanner.zip plug-in.
```
Comment (Tal-Daniel) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md:248]: link?

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+
+## Procedure
+1. On the collector, go to Setup > Tools and Views > Configure Universal Connector.
+2. Enable the connector if it is already disabled, before proceeding uploading of the UC.
+3. Click upload File and select the guardium_logstash-offline-plugin-gcp-pubsub-spanner.zip plug-in. After it is uploaded, click OK.
+4. Click Upload File and select the key.json file. After it is uploaded, click OK.
+5. Click the Plus sign to open the Connector Configuration dialog box.
+6. Type a name in the Connector name field.
+7. Update the input section to add the details from spanner_with_pubsub.conf file's input part, omitting the keyword "input{" at the beginning and its corresponding "}" at the end.
+8. Update the filter section to add the details from spanner_with_pubsub.conf file's filter part, omitting the keyword "filter{" at the beginning and its corresponding "}" at the end.
```
Comment (Tal-Daniel) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md:258]: see comments above. You might as well delete above sections with conf related comments, and use these sentences as starting point, just link to the conf you refer to in items 7 and 8.

@AlexaNahum

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+
+### Notes:
+* Replace "?" with this plugin version
+* logstash-plugin may not handle relative paths well, so try to install the gem from a simple path, as in the example above.
+## Run on local Logstash
+To test your filter using your local Logstash installation, run
+
+    $ ~/Downloads/logstash-7.14/bin/logstash -f filter-test.conf
+
+This configuration file generates an Event and send it thru the installed filter plugin.
```
Comment (Tal-Daniel) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md:300]: "... and send[s] it ..." @AlexaNahum

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/ExecuteSqlParser.java [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+
+import com.eclipsesource.v8.V8;
+import com.eclipsesource.v8.V8Object;
+import com.eclipsesource.v8.utils.V8ObjectUtils;
+
+public class ExecuteSqlParser {
+	
+	private static String customDDlRegex = "(\\w+\\s+(?i)table)\\s+([\\w]+)|(\\w+\\s+(?i)index)\\s+([\\w]+)|(\\w+\\s+(?i)view)\\s+([\\w]+)|(\\w+\\s+\\w+\\s+\\w+\\s+(?i)view)\\s+([\\w]+)|(\\w+\\s+(?i)database)\\s+([\\w-]+)";
+	private static String customSelectRegex = "(?i)FROM\\s*(?!\\s+(?i)UNNEST)\\s+(\\w+)";
+	private static String replaceDDLRegex = "((?i)UNIQUE)|((?i)NULL_FILTERED)";
```
Comment (Tal-Daniel) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/ExecuteSqlParser.java:32]: Regex patterns should have a comment next to them explaining what they do in plain English.

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/ExecuteSqlParser.java [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+			for (int i = 1; i <= matcher2.groupCount(); i++) {
+				if (matcher2.group(i) != null) {
+					lst.add(matcher2.group(i));
+				}
+			}
+		}
+		return lst;
+	}
+
+	public static Map<String, List<String>> parseSQL(V8Object parser, String sql) throws Exception {
```
Comment (Tal-Daniel) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/ExecuteSqlParser.java:118]: Could you add inline comments, or javadoc function comment, that explain reasoning behind multiple sections within function?

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/Parser.java [hunk 11]
```diff
@@ -0,0 +1,502 @@
...
+
+public class Parser {
+
+	private static Logger logger = LogManager.getLogger(Parser.class);
+	
+	/* this custom regex simply find the INTERSECT ALL/INTERSECT DISTINCT/EXCEPT ALL/EXCEPT DISTINCT keyword from the query.If it is found then we will perform logic accordingly*/
+	private static String customRegex = "((?i)\\s*intersect\\s+(?i)all)|((?i)\\s*INTERSECT(?i)\\s+DISTINCT)|((?i)\\s*except\\s+(?i)All)|((?i)\\s*EXCEPT\\s+(?i)DISTINCT)";
+	
+	/* this redactedRegex find the value within double quotes, single quotes and numeric values */
+	private static String redactedRegex = "\"[^\"]*\"|'[^']*'|\\b[\\d]+\\b";
```
Comment (Tal-Daniel) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/Parser.java:43]: Regex patterns should have a comment next to them explaining what they do in plain English.

**Review by Tal-Daniel — CHANGES_REQUESTED** [review]
See comments below.

**Comment by Tal-Daniel** [comment]
Sandeep, could you also add the resources the JS file before build/compilation?

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md (hunk 8 above)
Comment (AlexaNahum) [reply filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md:300]: Good catch. I caught a few more things now, here is the new proposed sentence:

This configuration file generates an event and sends it through the installed filter plug-in.

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/resources/LICENSE.txt [hunk 12]
```diff
@@ -0,0 +1,339 @@
+ GNU GENERAL PUBLIC LICENSE
+                       Version 2, June 1991
+
+ Copyright (C) 1989, 1991 Free Software Foundation, Inc.,
+ 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ Everyone is permitted to copy and distribute verbatim copies
+ of this license document, but changing it is not allowed.
```
Comment (Tal-Daniel) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/resources/LICENSE.txt:7]: Remove file, not necessary as mentioned in NOTICE and in file. 
Also, reminder: please include the JS file before build, to this folder. Thanks.

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md [hunk 13]
```diff
@@ -110,191 +115,36 @@ AND operation.producer="spanner.googleapis.com"
...
+8. Update the filter section to add the details from [spanner_with_pubsub.conf](spanner_with_pubsub.conf) file's filter part, omitting the keyword "filter{" at the beginning and its corresponding "}" at the end.
 9. 'type' field should match in input and filter configuration section.This field should be unique for every individual connector added.
 10. Click Save. Guardium validates the new connector, and enables the universal connector if it was disabled. After it is validated, it appears in the Configure Universal Connector page.
 11. After installing the plugins offline package and once the configuration is uploaded and saved to your Guardium machine, restart Universal Connector using the Disable/Enable button.
 12. Connect to the spanner instance and run queries.
 
-# Set-up dev environment
-Before you can build & create an updated GEM of this filter plugin, set up your environment as follows: 
-1. Clone Logstash codebase & build its libraries as as specified in [How to write a java filter plugin](https://github.com/logstash-plugins/logstash-filter-java_filter_example). Use branch 7.x (this filter was developed alongside 7.14 branch).
```
Comment (Tal-Daniel) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/README.md]: I rather like this section. It could really help developers feel guided, when they are introduced to the plug-in and want to try to install it, run it, fix an issue. Can we agree to return this section, to be after limitations, in all plug-ins? 
@ofer-haim @PratikshaSonawane @AlexaNahum

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/ExecuteSqlParser.java [hunk 14]
```diff
@@ -26,9 +26,23 @@
...
-	
+	/**
+	 * DDL queries are not supported Due to parser limitations,Hence handling in
+	 * regex to get object and verb values
+	 * 
+	 */
 	private static String customDDlRegex = "(\\w+\\s+(?i)table)\\s+([\\w]+)|(\\w+\\s+(?i)index)\\s+([\\w]+)|(\\w+\\s+(?i)view)\\s+([\\w]+)|(\\w+\\s+\\w+\\s+\\w+\\s+(?i)view)\\s+([\\w]+)|(\\w+\\s+(?i)database)\\s+([\\w-]+)";
+	/**
+	 * For Unnest function queries will only receive for verb values as not receive
+	 * object values
```
Comment (Tal-Daniel) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/ExecuteSqlParser.java:37]: Could you clarify? I did not understand.

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/ExecuteSqlParser.java [hunk 15]
```diff
@@ -26,9 +26,23 @@
 import com.eclipsesource.v8.utils.V8ObjectUtils;
 
 public class ExecuteSqlParser {
-	
+	/**
+	 * DDL queries are not supported Due to parser limitations,Hence handling in
+	 * regex to get object and verb values
+	 * 
+	 */
 	private static String customDDlRegex = "(\\w+\\s+(?i)table)\\s+([\\w]+)|(\\w+\\s+(?i)index)\\s+([\\w]+)|(\\w+\\s+(?i)view)\\s+([\\w]+)|(\\w+\\s+\\w+\\s+\\w+\\s+(?i)view)\\s+([\\w]+)|(\\w+\\s+(?i)database)\\s+([\\w-]+)";
```
Comment (Tal-Daniel) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/java/com/ibm/guardium/spanner/ExecuteSqlParser.java:34]: Could you help developers understand also what the regex is looking for, in general? for example, " ... to get object and verb values, by looking for words table, index, view, that appear _________ within the string."

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/resources/LICENSE.txt (hunk 12 above)
Comment (Tal-Daniel) [reply filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/resources/LICENSE.txt:7]: @sandeepkumarverma1 awaits removal & original JS file, if available.

**Review by Tal-Daniel — CHANGES_REQUESTED** [review]

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/resources/LICENSE.txt (hunk 12 above)
Comment (Tal-Daniel) [reply filter-plugin/logstash-filter-pubsub-spanner-guardium/src/main/resources/LICENSE.txt:7]: @sandeepkumarverma1 , instead of removing this file, please overwrite the plug-in LICENSE file with this one, as using GPL in a plug-in makes the whole plug-in under GPL license.

File: filter-plugin/logstash-filter-pubsub-spanner-guardium/NOTICE [hunk 16]
```diff
@@ -0,0 +1,5 @@
+filename - bigquery.peg-es5-v6.js
+
+--
+The file bigquery.peg-es5-v6.js was modified by IBM Corporation from bigquery.pegjs which is a part of node-sql-parser (https://github.com/taozhi8833998/node-sql-parser) package which comes under the GNU General Public License v2.0 (https://github.com/taozhi8833998/node-sql-parser/blob/master/LICENSE). bigquery.peg-es5-v6.js is licensed under the GNU General Public License v2.0.
```
Comment (Tal-Daniel) [inline filter-plugin/logstash-filter-pubsub-spanner-guardium/NOTICE:4]: Please add at the beginning of NOTICE that this plug-in is released under GPL, by IBM.

**Review by Tal-Daniel — CHANGES_REQUESTED** [review]
See previous comments, and these changes regarding LICENSE and NOTICE:

**Comment by sandeepkumarverma1** [comment]
This pull Request is very old, We have to build a new parser base. So dropping this PR as this is outdated now.
